File: 10. Mirror tree.cpp
#/*
 Task: Mirror tree (swap left/right subtrees)
 Time Complexity: O(n)
 Space Complexity: O(h) — recursion stack
*/
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
void mirror(Node* root)
{
    if (!root)
        return;
    swap(root->left, root->right);
    mirror(root->left);
    mirror(root->right);
}
void inorder(Node* root)
{
    if (!root)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    cout << "Before inorder: ";
    inorder(root);
    cout << "\n";
    mirror(root);
    cout << "After inorder: ";
    inorder(root);
    cout << endl;
    return 0;
}
// sc tc

File: 10. Mirror tree.py
"""
Task: Mirror tree
Time Complexity: O(n)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def mirror(root):
    if not root: return
    root.left, root.right = root.right, root.left
    mirror(root.left); mirror(root.right)

def inorder(root):
    if not root: return
    inorder(root.left); print(root.data, end=' '); inorder(root.right)

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4)
    print('Before inorder:', end=' '); inorder(root); print()
    mirror(root)
    print('After inorder:', end=' '); inorder(root); print()
